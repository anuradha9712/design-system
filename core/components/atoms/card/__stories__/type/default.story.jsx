import * as React from 'react';
import { Card } from '@/index';

// CSF format story
export const defaultCard = () => {
  return (
    <Card>
      <div className="p-8" />
    </Card>
  );
};

export default {
  title: 'Layout/Card/Type/Default Card',
  component: Card,
};
