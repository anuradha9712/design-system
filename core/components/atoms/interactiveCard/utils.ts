export const removeItem = (targetItem: any, arr = []) => {
  return arr.filter((item: any) => item !== targetItem);
};
