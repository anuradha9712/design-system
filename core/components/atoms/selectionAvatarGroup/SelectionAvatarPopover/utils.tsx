import React from 'react';

export const handleKeyDown = (
  event: React.KeyboardEvent,
  focusedOption: Element | undefined,
  setFocusedOption?: React.Dispatch<React.SetStateAction<Element | undefined>>,
  setHighlightFirstItem?: React.Dispatch<React.SetStateAction<boolean>>,
  setHighlightLastItem?: React.Dispatch<React.SetStateAction<boolean>>,
  listRef?: any,
  withSearch?: boolean
) => {
  switch (event.key) {
    case 'ArrowUp':
      event.preventDefault();
      navigateOptions('up', focusedOption, setFocusedOption, listRef, withSearch);
      break;
    case 'ArrowDown':
      event.preventDefault();
      navigateOptions('down', focusedOption, setFocusedOption, listRef, withSearch);
      break;
    case 'Enter':
      handleEnterKey(focusedOption);
      setHighlightLastItem?.(false);
      setHighlightFirstItem?.(false);
      break;
    default:
      break;
  }
};

const handleEnterKey = (focusedOption: Element | undefined) => {
  (focusedOption as HTMLElement)?.click();
};

const navigateSelectedOption = (
  index: number,
  withSearch?: boolean,
  listItems?: any,
  direction?: string,
  listRef?: any,
  setFocusedOption?: React.Dispatch<React.SetStateAction<Element | undefined>>
) => {
  console.log('tttindexxx', index);
  // if (withSearch) {
  //   const searchInput = listRef.current.querySelector('[data-test="DesignSystem-SelectionAvatar--Input"]');
  //   console.log('ttt inside iff');
  //   if ((index === 0 && direction === 'up') || (index === listItems.length - 2 && direction === 'down')) {
  //     console.log('tttfocus search');
  //     searchInput.focus();
  //     setFocusedOption && setFocusedOption(searchInput);
  //     return;
  //   }
  // }
  const searchInput = listRef.current.querySelector('[data-test="DesignSystem-SelectionAvatar--Input"]');
  searchInput.focus();
  setFocusedOption && setFocusedOption(searchInput);
};

const navigateOptions = (
  direction: string,
  focusedOption: Element | undefined,
  setFocusedOption?: React.Dispatch<React.SetStateAction<Element | undefined>>,
  listRef?: any,
  withSearch?: boolean
) => {
  const listItems = listRef.current.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
  let index = Array.from(listItems).findIndex((item) => {
    return item == focusedOption;
  });

  console.log('tttwithSearchwithSearch', withSearch, 'index', index, 'listItems[index]', listItems[index]);

  if (index === -1) {
    console.log('ttt11');
    index = direction === 'up' ? listItems.length - 1 : 0;
  } else if (
    (withSearch && index === 0 && direction === 'up') ||
    (withSearch && index === listItems.length - 1 && direction === 'down')
  ) {
    console.log('ttt22');

    navigateSelectedOption(index, withSearch, listItems, direction, listRef, setFocusedOption);
    // return;
  } else {
    console.log('ttt333');

    index = direction === 'up' ? (index - 1 + listItems.length) % listItems.length : (index + 1) % listItems.length;

    const targetOption = listItems[index];

    (targetOption as HTMLElement).focus();
    setFocusedOption && setFocusedOption(targetOption);
    targetOption.scrollIntoView({ block: 'center' });
  }
};

export const handleInputKeyDown = (
  event: React.KeyboardEvent,
  listRef: any,
  setFocusedOption: React.Dispatch<React.SetStateAction<Element | undefined>>
) => {
  const listItems = listRef.current?.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
  let targetOption;

  switch (event.key) {
    case 'ArrowUp':
      event.preventDefault();
      targetOption = listItems[listItems.length - 1];
      break;
    case 'ArrowDown':
      event.preventDefault();
      targetOption = listItems[0];
      break;
    default:
      break;
  }

  (targetOption as HTMLElement)?.focus();
  targetOption?.scrollIntoView({ block: 'center' });
  setFocusedOption && setFocusedOption(targetOption);
};
