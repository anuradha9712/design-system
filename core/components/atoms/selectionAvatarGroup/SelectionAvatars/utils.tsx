import React from 'react';

export const handleKeyDown = (
  event: React.KeyboardEvent,
  setOpenPopover?: React.Dispatch<React.SetStateAction<boolean>>,
  setHighlightFirstItem?: React.Dispatch<React.SetStateAction<boolean>>,
  setHighlightLastItem?: React.Dispatch<React.SetStateAction<boolean>>
) => {
  switch (event.key) {
    case 'Enter':
      setOpenPopover?.(true);
      setHighlightFirstItem?.(true);
      break;
    case 'ArrowDown':
      setOpenPopover?.(true);
      setHighlightFirstItem?.(true);
      break;
    case 'ArrowUp':
      setOpenPopover?.(true);
      setHighlightLastItem?.(true);
      break;
    default:
      break;
  }
};

export const focusListItem = (
  position: string,
  setFocusedOption?: React.Dispatch<React.SetStateAction<Element | undefined>>,
  listRef?: any,
  withSearch?: boolean
) => {
  if (withSearch && position === 'down') {
    const searchInput = listRef.current?.querySelectorAll('[data-test="DesignSystem-SelectionAvatar--Input"]');
    console.log('aa22', searchInput);
    searchInput[0]?.focus();
    setFocusedOption && setFocusedOption(searchInput);
    return;
  }

  const listItems = listRef.current?.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
  console.log('listReflistRef', listRef.current, 'listItems', listItems);
  let targetOption;

  if (position === 'down') {
    targetOption = listItems[0];
  } else {
    targetOption = listItems[listItems.length - 1];
  }
  (targetOption as HTMLElement)?.focus();
  targetOption?.scrollIntoView({ block: 'center' });
  setFocusedOption && setFocusedOption(targetOption);
};
