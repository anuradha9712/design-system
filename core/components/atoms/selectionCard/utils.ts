export const removeItem = (targetItem: string, arr: string[]) => {
  return arr.filter((item: string) => item !== targetItem);
};

export const handleMultiSelect = (
  selectedOptions: string[],
  isItemSelected: boolean,
  setSelectedOptions: React.Dispatch<React.SetStateAction<string[]>>,
  id: string
) => {
  let resultList = [...selectedOptions];
  if (isItemSelected) {
    resultList = removeItem(id, selectedOptions);
  } else {
    resultList.push(id);
  }
  setSelectedOptions(resultList);
};

export const handleSingleSelect = (
  selectedOptions: string[],
  isItemSelected: boolean,
  setSelectedOptions: React.Dispatch<React.SetStateAction<string[]>>,
  id: string
) => {
  let resultList = [...selectedOptions];
  if (isItemSelected) {
    resultList = [];
  } else {
    resultList = [id];
  }

  setSelectedOptions(resultList);
};
