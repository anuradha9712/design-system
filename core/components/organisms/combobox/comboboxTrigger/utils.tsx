import React from 'react';

export const handleKeyDown = (
  event: React.KeyboardEvent,
  setOpenPopover?: React.Dispatch<React.SetStateAction<boolean>>,
  // setFocusedOption?: React.Dispatch<React.SetStateAction<Element | undefined>>,
  setHighlightFirstItem?: any,
  setHighlightLastItem?: any
) => {
  console.log('handle key down e.key', event.key);

  switch (event.key) {
    case 'ArrowUp':
      setOpenPopover?.(true);
      setHighlightLastItem(true);
      // focusListItem('up', setFocusedOption);
      // requestAnimationFrame(() => focusListItem('up', setFocusedOption));
      break;
    case 'ArrowDown':
      setOpenPopover?.(true);
      setHighlightFirstItem(true);

      // requestAnimationFrame(() => focusListItem('down', setFocusedOption));
      break;
    case 'Escape':
      setOpenPopover?.(false);
      break;
    default:
      break;
  }
};

export const focusListItem = (
  position: string,
  setFocusedOption?: React.Dispatch<React.SetStateAction<Element | undefined>>
) => {
  const listItems = document.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
  let targetOption;
  // if (position === 'down') {
  //   targetOption = listItems[0];
  //   (listItems[0] as HTMLElement)?.focus();
  // } else {
  //   (listItems[listItems.length - 1] as HTMLElement)?.focus();
  // }

  if (position === 'down') {
    console.log('rrr down', listItems);
    targetOption = listItems[0];
  } else {
    targetOption = listItems[listItems.length - 1];
  }
  console.log('rrrtargetOption', targetOption);
  (targetOption as HTMLElement)?.focus();
  targetOption?.scrollIntoView({ block: 'center' });
  setFocusedOption && setFocusedOption(targetOption);
};
