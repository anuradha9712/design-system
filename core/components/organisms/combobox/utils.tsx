import React from 'react';

export const handleKeyDown = (
  event: React.KeyboardEvent,
  focusedOption: Element | undefined,
  setFocusedOption?: React.Dispatch<React.SetStateAction<Element | undefined>>,
  setOpenPopover?: React.Dispatch<React.SetStateAction<boolean>>,
  inputTriggerRef?: any
) => {
  switch (event.key) {
    case 'ArrowUp':
      event.preventDefault();
      navigateOptions('up', focusedOption, setFocusedOption);
      break;
    case 'ArrowDown':
      event.preventDefault();
      navigateOptions('down', focusedOption, setFocusedOption);
      break;
    case 'Enter':
      handleEnterKey(focusedOption);
      break;
    case 'Escape':
      setOpenPopover?.(false);
      console.log('inputTriggerRef.current', inputTriggerRef.current);
      inputTriggerRef.current.focus();
      break;
    default:
      break;
  }
};

const handleEnterKey = (focusedOption: Element | undefined) => {
  (focusedOption as HTMLElement)?.click();
};

const navigateOptions = (
  direction: string,
  focusedOption: Element | undefined,
  setFocusedOption?: React.Dispatch<React.SetStateAction<Element | undefined>>
) => {
  const listItems = document.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
  let index = Array.from(listItems).findIndex((item) => {
    return item == focusedOption;
  });

  if (index === -1) {
    index = direction === 'up' ? listItems.length - 1 : 0;
  } else {
    index = direction === 'up' ? (index - 1 + listItems.length) % listItems.length : (index + 1) % listItems.length;
  }

  const targetOption = listItems[index];
  console.log('targetOptiontargetOption', targetOption);
  (targetOption as HTMLElement).focus();
  setFocusedOption && setFocusedOption(targetOption);
  targetOption.scrollIntoView({ block: 'center' });
};
