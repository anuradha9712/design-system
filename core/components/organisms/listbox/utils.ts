const focusOption = (targetIndex: number, totalLength: number, direction: string) => {
  let startIndex = targetIndex;
  const endIndex = direction === 'down' ? totalLength : -1;

  while (startIndex !== endIndex) {
    const selectedElement = document.querySelector(`[data-list-item-index="${startIndex}"]`) as HTMLElement;

    if (selectedElement && selectedElement.getAttribute('data-disabled') !== 'true') {
      selectedElement && selectedElement.focus();
      break;
    }

    if (direction === 'down') {
      startIndex++;
    } else {
      startIndex--;
    }
  }
};

export const onKeyDown = (event: React.KeyboardEvent, totalLength: number) => {
  const sourceElement = event.target as HTMLElement;
  const sourceIndex = sourceElement.getAttribute('data-list-item-index');
  const targetIndex = (sourceIndex && parseInt(sourceIndex, 10)) || 0;

  switch (event.key) {
    case 'ArrowDown':
      event.preventDefault();
      focusOption(targetIndex + 1, totalLength, 'down');
      break;
    case 'ArrowUp':
      event.preventDefault();
      focusOption(targetIndex - 1, totalLength, 'up');
      break;
    case 'Enter': {
      event.preventDefault();
      sourceElement.click();
      break;
    }
    default:
      break;
  }
};
