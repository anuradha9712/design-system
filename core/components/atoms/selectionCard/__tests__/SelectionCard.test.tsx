import * as React from 'react';
import { fireEvent, render } from '@testing-library/react';
import { testHelper, filterUndefined, valueHelper, testMessageHelper } from '@/utils/testHelper';
import { SelectionCard } from '@/index';
import { SelectionCardProps as Props, CardItemProps } from '@/index.type';

const BooleanValue = [true, false];
const FunctionValue = jest.fn();
const StringValue = 'Selection Card';
const ID = 'card1';

describe('selection card component snapshot', () => {
  const mapper: Record<string, any> = {
    multiSelect: valueHelper(BooleanValue, { required: false, iterate: true }),
  };

  const testFunc = (props: Record<string, any>): void => {
    const attr = filterUndefined(props) as Props;

    it(testMessageHelper(attr), () => {
      const { baseElement } = render(<SelectionCard {...attr} />);
      expect(baseElement).toMatchSnapshot();
    });
  };

  testHelper(mapper, testFunc);
});

describe('selection card item component snapshot', () => {
  const mapper: Record<string, any> = {
    disabled: valueHelper(BooleanValue, { required: false, iterate: true }),
    children: valueHelper(StringValue, { required: true }),
    onClick: valueHelper(FunctionValue, { required: false }),
  };

  const testFunc = (props: Record<string, any>): void => {
    const attr = filterUndefined(props) as CardItemProps;

    it(testMessageHelper(attr), () => {
      const { baseElement } = render(<SelectionCard.Item {...attr} />);
      expect(baseElement).toMatchSnapshot();
    });
  };

  testHelper(mapper, testFunc);
});

/*
onClick-> controlled
*/

describe('selection card item component tests', () => {
  it('check for render item', () => {
    const { getByTestId } = render(<SelectionCard.Item id={ID}>{StringValue}</SelectionCard.Item>);
    const element = getByTestId('DesignSystem-SelectionCard-Item');
    expect(element).toBeInTheDocument();
    expect(element).toHaveClass('Selection-card');
    expect(element).toHaveTextContent(StringValue);
  });
});

describe('selection card item overlay component tests', () => {
  it('check for overlay item', () => {
    const { getByTestId } = render(<SelectionCard.Item id={ID}>{StringValue}</SelectionCard.Item>);
    const element = getByTestId('DesignSystem-SelectionCard-Overlay');
    expect(element).toBeInTheDocument();
    expect(element).toHaveClass('Selection-card-overlay');
  });

  it('check for overlay item z-index', () => {
    const { getByTestId } = render(
      <SelectionCard.Item id={ID} zIndex={100}>
        {StringValue}
      </SelectionCard.Item>
    );
    const element = getByTestId('DesignSystem-SelectionCard-Overlay');
    expect(element).toHaveStyle({ zIndex: 100 });
  });
});

describe('selection card disabled item component tests', () => {
  it('check for disabled class for unselected item', () => {
    const { getByTestId } = render(
      <SelectionCard.Item id={ID} disabled={true}>
        {StringValue}
      </SelectionCard.Item>
    );
    const element = getByTestId('DesignSystem-SelectionCard-Item');
    expect(element).toHaveClass('Selection-card--disabled');
  });

  it('check for disabled class for selected item', () => {
    const { getByTestId } = render(
      <SelectionCard selectedList={[ID]}>
        <SelectionCard.Item id={ID} disabled={true}>
          {StringValue}
        </SelectionCard.Item>
      </SelectionCard>
    );
    const element = getByTestId('DesignSystem-SelectionCard-Item');
    expect(element).toHaveClass('Selection-card--selected-disabled');
  });
});

describe('selection card component classes', () => {
  it('check for selection classes', () => {
    const { getByTestId } = render(
      <SelectionCard selectedList={[ID]}>
        <SelectionCard.Item id={ID}>{StringValue}</SelectionCard.Item>
      </SelectionCard>
    );
    const element = getByTestId('DesignSystem-SelectionCard-Item');
    expect(element).toHaveClass('Selection-card--selected');
  });
});

describe('selection card component multiselect', () => {
  it('check for selection classes for multiselect:false', () => {
    const { getAllByTestId } = render(
      <SelectionCard selectedList={['card1']}>
        <SelectionCard.Item id={'card1'}>{StringValue}</SelectionCard.Item>
        <SelectionCard.Item id={'card2'}>{StringValue}</SelectionCard.Item>
      </SelectionCard>
    );
    const element = getAllByTestId('DesignSystem-SelectionCard-Item');
    expect(element[0]).toHaveClass('Selection-card--selected');
    fireEvent.click(element[1]);
    expect(element[0]).not.toHaveClass('Selection-card--selected');
    expect(element[1]).toHaveClass('Selection-card--selected');
  });

  it('check for selection classes for multiselect:false', () => {
    const { getAllByTestId } = render(
      <SelectionCard selectedList={['card1']} multiSelect={true}>
        <SelectionCard.Item id={'card1'}>{StringValue}</SelectionCard.Item>
        <SelectionCard.Item id={'card2'}>{StringValue}</SelectionCard.Item>
      </SelectionCard>
    );
    const element = getAllByTestId('DesignSystem-SelectionCard-Item');
    expect(element[0]).toHaveClass('Selection-card--selected');
    fireEvent.click(element[1]);
    expect(element[0]).toHaveClass('Selection-card--selected');
    expect(element[1]).toHaveClass('Selection-card--selected');
  });
});

describe('selection card component event handlers', () => {
  it('check for click event handler', () => {
    const { getByTestId } = render(
      <SelectionCard selectedList={[ID]}>
        <SelectionCard.Item id={ID} onClick={FunctionValue}>
          {StringValue}
        </SelectionCard.Item>
      </SelectionCard>
    );
    const element = getByTestId('DesignSystem-SelectionCard-Item');
    fireEvent.click(element);
    expect(FunctionValue).toHaveBeenCalled();
    expect(FunctionValue).toHaveBeenCalledWith(expect.objectContaining({}), ID);
  });
});

describe('selection card component custom classes', () => {
  it('check for custom class in selection card component', () => {
    const { getByTestId } = render(<SelectionCard className="custom-class">{StringValue}</SelectionCard>);
    expect(getByTestId('DesignSystem-SelectionCard-OuterWrapper')).toHaveClass('custom-class');
  });

  it('check for custom class in selection card item component', () => {
    const { getByTestId } = render(
      <SelectionCard.Item className="custom-class" id={ID}>
        {StringValue}
      </SelectionCard.Item>
    );
    expect(getByTestId('DesignSystem-SelectionCard-Item')).toHaveClass('custom-class');
  });
});
