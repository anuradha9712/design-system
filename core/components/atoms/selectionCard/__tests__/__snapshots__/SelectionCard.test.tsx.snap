// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selection card component snapshot 
  
 1`] = `
<body>
  <div>
    <div
      data-test="DesignSystem-SelectionCard-OuterWrapper"
    />
  </div>
</body>
`;

exports[`selection card component snapshot 
  multiSelect: false
 1`] = `
<body>
  <div>
    <div
      data-test="DesignSystem-SelectionCard-OuterWrapper"
    />
  </div>
</body>
`;

exports[`selection card component snapshot 
  multiSelect: true
 1`] = `
<body>
  <div>
    <div
      data-test="DesignSystem-SelectionCard-OuterWrapper"
    />
  </div>
</body>
`;

exports[`selection card item component snapshot 
  children: "Selection Card", disabled: false, onClick: [null]
 1`] = `
<body>
  <div>
    <div
      class="Selection-card"
      data-test="DesignSystem-SelectionCard-Item"
      role="button"
      tabindex="0"
    >
      <div
        class="Selection-card-overlay"
        data-test="DesignSystem-SelectionCard-Overlay"
        style="z-index: 2;"
      />
      Selection Card
    </div>
  </div>
</body>
`;

exports[`selection card item component snapshot 
  children: "Selection Card", disabled: true, onClick: [null]
 1`] = `
<body>
  <div>
    <div
      class="Selection-card Selection-card--disabled"
      data-test="DesignSystem-SelectionCard-Item"
      role="button"
      tabindex="0"
    >
      <div
        class="Selection-card-overlay"
        data-test="DesignSystem-SelectionCard-Overlay"
        style="z-index: 2;"
      />
      Selection Card
    </div>
  </div>
</body>
`;

exports[`selection card item component snapshot 
  children: "Selection Card", onClick: [null]
 1`] = `
<body>
  <div>
    <div
      class="Selection-card"
      data-test="DesignSystem-SelectionCard-Item"
      role="button"
      tabindex="0"
    >
      <div
        class="Selection-card-overlay"
        data-test="DesignSystem-SelectionCard-Overlay"
        style="z-index: 2;"
      />
      Selection Card
    </div>
  </div>
</body>
`;
